// @huddle01/server-sdk/dist/auth/AccessToken.cjs
'use strict';

var chunkMCJQHYW7_cjs = require('../chunk-MCJQHYW7.cjs');
require('../chunk-UYXCM27V.cjs');
require('../chunk-5G3D25PZ.cjs');



Object.defineProperty(exports, "AccessToken", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.AccessToken; }
});
Object.defineProperty(exports, "DEFAULT_PERMISSIONS", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.DEFAULT_PERMISSIONS; }
});
Object.defineProperty(exports, "Permissions", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.Permissions; }
});
Object.defineProperty(exports, "ROLE_PERMISSIONS", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.ROLE_PERMISSIONS; }
});
Object.defineProperty(exports, "Role", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.Role; }
});
Object.defineProperty(exports, "RoleSchema", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.RoleSchema; }
});
Object.defineProperty(exports, "TokenVerifier", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.TokenVerifier; }
});
Object.defineProperty(exports, "isDefaultRole", {
  enumerable: true,
  get: function () { return chunkMCJQHYW7_cjs.isDefaultRole; }
});
