// @huddle01/server-sdk/dist/auth/AccessToken.cjs
'use strict';

var chunk5DEEA3ZI_cjs = require('../chunk-5DEEA3ZI.cjs');
require('../chunk-NJUEGIAO.cjs');
require('../chunk-UYXCM27V.cjs');
require('../chunk-VIECOXGY.cjs');



Object.defineProperty(exports, "AccessToken", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.AccessToken; }
});
Object.defineProperty(exports, "DEFAULT_PERMISSIONS", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.DEFAULT_PERMISSIONS; }
});
Object.defineProperty(exports, "MeetingTokenInputSchema", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.MeetingTokenInputSchema; }
});
Object.defineProperty(exports, "Permissions", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.Permissions; }
});
Object.defineProperty(exports, "ROLE_PERMISSIONS", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.ROLE_PERMISSIONS; }
});
Object.defineProperty(exports, "Role", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.Role; }
});
Object.defineProperty(exports, "RoleSchema", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.RoleSchema; }
});
Object.defineProperty(exports, "TokenVerifier", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.TokenVerifier; }
});
Object.defineProperty(exports, "isDefaultRole", {
  enumerable: true,
  get: function () { return chunk5DEEA3ZI_cjs.isDefaultRole; }
});
